src/app/berkeley_migration_verifier/berkeley_migration_verifier.ml

```diff
@@ -0,0 +1,389 @@
+    query_migrated_db ~f:(fun db -> Sql.Berkeley.count_orphaned_blocks db)
+  in
+  let%bind orphaned_blocks =
+    query_migrated_db ~f:(fun db -> Sql.Berkeley.count_pending_blocks db)
+  in
+  match orphaned_blocks + pending_blocks with
+  | 0 ->
+      Deferred.return None
+  | _ ->
+      Deferred.return
+        (Some
+           (sprintf
+              "Expected to have at only canonical block while having %d \
+               orphaned and %d pending"
+              orphaned_blocks pending_blocks ) )
+
+let printout_errors_or_success error_messages ~check_name =
+  let print_errors error_messages =
+    if List.length error_messages > 10 then (
+      Async.printf " - Details (truncated to only 10 errors): " ;
+      List.iter (List.take error_messages 10) ~f:(fun error_message ->
+          printf "\t%s\n" error_message ) )
+    else printf " - Details : " ;
+    List.iter error_messages ~f:(fun error_message ->
+        printf "\t%s\n" error_message )
+  in
+  if List.length error_messages > 1 then (
+    exit_code := !exit_code + 1 ;
+    printf "%s check ... ❌ \n" check_name ;
+    print_errors error_messages )
+  else printf "%s check ... ✅ \n" check_name
+
+let compare_db_content ~mainnet_archive_uri ~migrated_archive_uri
+    ~fork_state_hash () =
+  printf
+    "Running verifications between '%s' and '%s' schemas. It may take a couple \
+     of minutes... \n"
+    mainnet_archive_uri migrated_archive_uri ;
+
+  let mainnet_archive_uri = Uri.of_string mainnet_archive_uri in
+  let migrated_archive_uri = Uri.of_string migrated_archive_uri in
+  let mainnet_pool =
+    Caqti_async.connect_pool ~max_size:128 mainnet_archive_uri
+  in
+  let migrated_pool =
+    Caqti_async.connect_pool ~max_size:128 migrated_archive_uri
+  in
+
+  match (mainnet_pool, migrated_pool) with
+  | Error _e, _ | _, Error _e ->
+      failwith " Failed to create Caqti pools for Postgresql ❌. Exiting \n"
+  | Ok mainnet_pool, Ok migrated_pool ->
+      printf "[1/9] Connection to orignal and migrated schema succesful ✅ \n" ;
+
+      let query_mainnet_db = Mina_caqti.query mainnet_pool in
+      let query_migrated_db = Mina_caqti.query migrated_pool in
+      let%bind end_global_slot =
+        get_migration_end_slot_for_state_hash ~query_mainnet_db fork_state_hash
+      in
+      let compare_hashes ~fetch_data_sql ~find_element_sql ~name =
+        let%bind expected_hashes = query_mainnet_db ~f:fetch_data_sql in
+        Deferred.List.filter_map expected_hashes ~f:(fun hash ->
+            let%bind element_id = find_element_sql hash in
+            if element_id |> Option.is_none then
+              return
+                (Some
+                   (sprintf "Cannot find %s hash ('%s') in migrated database"
+                      name hash ) )
+            else return None )
+      in
+
+      let%bind error_messages =
+        compare_hashes
+          ~fetch_data_sql:(fun db ->
+            Sql.Mainnet.user_commands_hashes db end_global_slot )
```

Comment:
This would only success if all blocks in mainnet database are canonical. If we have pending block that contains some user command which is not included in the canonical blocks, then this check would fail. Maybe it's better to make sure the block is canonical when you query anything mainnet.